keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+    `Rule` expression should always evaluate a boolean. Token `claims` are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
+
+    You can find a snippet of validation rules below:
+
+    ```yaml
+    claimValidationRules:
+    
+    - rule: 'claims.aud == "charmander" || claims.aud == "bulbasaur"'
+      message: clients other than charmander or bulbasaur are not allowed
+    
+    - rule: 'claims.roles.exists(r, r == "kubernetes-user")'
+      message: only kubernetes-user group members can access the cluster
+    ```
+
+3. `ClaimMappings` - rules to map claims from a token to Kubernetes user attributes.
+    ```go
+    type UserAttributes struct {
+        // Username represents an option for the username attribute.
+        Username string
+        // Groups represents an option for the groups attribute. 
+        // +optional
+        Groups string
+        // UID represents an option for the uid attribute.
+        // +optional
+        UID string
+        // Extra represents an option for the extra attribute.
+        // +optional
+        Extra []ExtraMapping
+    }
+   
+    type ExtraMapping struct {
+        // Key is a CEL expression to extract extra attribute key.
+        Key string
+        // Expression is a CEL expression to extract extra attribute value.
+        Expression string
+    }
+    ```
+
+    Every field of structures above is a CEL expression. For username and uid, the expression should evaluate a string.
+    For groups, it should evaluate to an array. A map of extra attributes is composed of key/value pairs, each with their
+    own CEL expressions, both should evaluate to a string.
+
+    The example of mapping user attributes can be found below:
+
+    ```yaml
+    claimMappings:
+      username: 'claims.username + ":external-user"'
+      groups: 'claims.roles.split(",")'
+      uid: 'claims.sub'
+      extra:
+      - key: '"client_name"'
+        expression: 'claims.aud'
+    ```
+   
+    For the token with the following claims
+    ```json
+    {
+      "sub": "119abc",
+      "aud": "kubernetes",
+      "username": "jane_doe",
+      "roles": "admin,user",
+      ...
+    }
+    ```
+    our authenticator will extract the following user attributes:
+    ```yaml
+    username: jane_doe:external-user
+    groups: ["admin", "user"]
+    uid: "119abc"
+    extra:
+      client_name: kubernetes
+    ```
+4. `ClaimsFilter` - list of claim names that should be passed to CEL expressions. The assumption is that administrators
+   know the structure of the token and the exact claims they will use in CEL expressions.
+   This option helps to reduce system load and operate only with required claims.
+
+### More about CEL
+
+* CEL runtime should be compiled only once if structured OIDC config option is enabled
```

Comment:
This is the first use of CEL in configuration files. So I'm thinking this through.

In CRDs, the CEL expressions are contained in the exact schema the expressions validate, so we fully compile CEL expressions, this includes full type-checking against the schema.

For ValidationAdmissionPolicies, the CEL expressions are defined in resources and are not tightly bound to the schemas of resources. Since, CRDs schemas can change at any time, and even for builtin types, version skew and resource portability between clusters make static type checking problematic. So we do not "compile" CEL expressions. Instead, we defer all type checking to runtime, this allows for expressions with CEL `has()` checks to be used to guard access to fields that might not _exist_. These expressions would fail type checking but can evaluate safely _if_ written carefully.  We are working on "informational type checking" to allow expression authors to observe type errors, but they won't be blocking.

For configuration files, since it is normal for flags to be added in new k8s versions, I think we need to take the same approach as for ValidationAdmissionPolicies. This would make it safe to rollback a component even after adding new flags.  Does this sound right?

cc @liggitt for how configuration files handle version skew.